### PR TITLE
test(e2e): add inactive Windows MXC OpenClaw qualification

### PR DESCRIPTION
**Pull request by senthilr-nv**
## Summary

Adds an opt-in live E2E target for the inactive Windows MXC OpenClaw `process_container` candidate. NemoClaw previously had no revision-bound qualification path for this candidate; the target now checks exact package and artifact identities, workload readiness, filesystem allow and deny behavior, sandbox deletion, registry removal, and exact OpenClaw process cleanup without registering or activating MXC.

## Related Issue

Refs #8178

## Changes

- Validate the exact NemoClaw revision, OpenShell package binaries, `wxc-exec`, and OpenClaw artifact tree before execution, then revalidate each pinned runtime artifact immediately before its security-relevant use.
- Generate a temporary least-privilege gateway configuration, filesystem policy, and OpenClaw probe, then exercise sandbox create and delete only through OpenShell.
- Restrict host child environments, fail closed on command/query errors, validate process identities before emergency cleanup, and record a secret-free qualification receipt.
- Add support tests, mock-parity coverage, and semantic checks for progress and the OpenShell-only MXC control boundary.

## Type of Change

- [ ] Code change (feature, bug fix, or refactor)
- [x] Code change with doc updates
- [ ] Doc only (prose changes, no code sample modifications)
- [ ] Doc only (includes code sample changes)

## Quality Gates

- [x] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [x] Docs updated for user-facing behavior changes
- [ ] Docs not applicable — justification:
- [x] Sensitive paths changed (security, policy, credentials, preflight, onboarding, inference, runner, sandbox, or messaging)
- [x] Sensitive-path review completed or maintainer-approved waiver recorded — reviewer/approval link/justification: [Exact-head security review](https://github.com/NVIDIA/NemoClaw/pull/8300#pullrequestreview-4866742249) passed all nine categories on `99e54157678400cfda425bdc9a1f0e531749e6ea`; the target is opt-in, loopback-bound, argument-array based, least-privilege, identity-pinned and revalidated before security-relevant use, host-environment restricted, token-redacting, resource-bounded, and fail-closed on create, query, or cleanup fallback.
- [ ] Non-success, skipped, or missing CI check accepted by maintainer — check name, approval link, and follow-up issue:

## Documentation Writer Review

- [x] Documentation writer subagent reviewed the completed changes
- Result: `pass`
- Evidence: Exact-head review confirmed the inactive/no-support boundary, artifact revalidation, least-privilege receipt, environment allowlist, process-identity checks, and failure-containment cleanup behavior. The writer corrected the receipt claim so operators are told that, after preflight and local setup succeed, the target writes a secret-free receipt for either verdict; earlier failures may occur before receipt creation. No user docs, changelog, navigation, or routing change is required.
- Agent: Codex Desktop documentation-writer subagent

## DGX Station Hardware Evidence

- [ ] Tested on DGX Station
- Tested commit:
- Station profile/scenario:
- Result:
- Supporting evidence:

## Verification

- [x] PR description includes a `Signed-off-by:` line and all 8 commits appear as `Verified` in GitHub
- [x] Normal `pre-commit`, `commit-msg`, and `pre-push` hooks passed, or `npm run validate:pr` passed after refreshing `origin/main` when hooks were skipped or unavailable
- [x] Targeted behavior tests pass for the current change set, or tests are marked not applicable above — `npx vitest run --project e2e-support test/e2e/support/windows-mxc-openclaw-process-container.test.ts test/e2e/support/e2e-semantic-phase-check.test.ts` passed 35 tests in 2 files; `npm run test:e2e-phases:check` passed 118 tests in 75 files; mock parity, repository checks, CLI build, and CLI type-check passed. The receipt-boundary follow-up passed normal Markdown lint and the repository test-file budget.
- [ ] Applicable broad gate passed — `npm test` for broad runtime/test-harness changes; `npm run check` for repo-wide validation/coverage changes — command/result: `npm test` was attempted outside the sandbox, including a four-worker retry, but unchanged local tests failed or timed out; an isolated failure was caused by the host Python lacking `yaml`.
- [x] Quality Gates section completed with required justifications or waivers
- [x] No secrets, API keys, or credentials committed
- [ ] `npm run docs` builds without warnings (doc changes only) — the build passed with two existing Fern warnings.
- [x] Doc pages follow the [style guide](https://github.com/NVIDIA/NemoClaw/blob/main/docs/CONTRIBUTING.md) (doc changes only)
- [ ] New doc pages include SPDX header and frontmatter (new pages only)

## GitHub Actions

Fresh exact-head CI and protected E2E evidence for `99e54157678400cfda425bdc9a1f0e531749e6ea` are required. No prior-head result is carried forward and no check is waived.

---
Signed-off-by: Senthil Ravichandran <senthilr@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Added qualification coverage for Windows x64 OpenClaw process containers, including sandbox lifecycle, process identity, artifact integrity, readiness, and cleanup validation.
  * Added deterministic artifact-tree integrity verification for qualification assets.

* **Tests**
  * Added gated end-to-end validation with progress reporting and pass/fail receipts.
  * Added contract and semantic checks for configuration, security boundaries, token handling, process validation, cleanup, and failure scenarios.

* **Documentation**
  * Documented prerequisites, environment setup, execution steps, artifact verification, token handling, cleanup safeguards, and expected failure behavior.

### CODE REVIEW

**Comment by cv** [comment]
Maintainer train classification for `adb7d8485a32071ad7bf1524a3330d6f63b7b204`:

This PR is within #8178 delivery step 5 because it adds an inactive, opt-in qualification harness and does not register, activate, or document Windows/MXC support. It is not merge-ready yet:

- Required CI fails because the new live target has no entry in `test/e2e/mock-parity.json`.
- Fail closed on an unrelated nonzero `sandbox create` result; accept an exception only for an exact documented OpenShell outcome protected by a regression test.
- Keep `pc_least_privilege` enabled, or document an exact version-bound upstream constraint and test the compensating privilege boundary.
- Add direct artifact-digest tests for content and relative-path changes plus symlink and unsupported-entry rejection.

These are technical and security validation gaps, not approval to broaden the product surface. No check waiver or activation is proposed.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 1]
```diff
@@ -0,0 +1,1375 @@
...
+        };
+        if (!registryRemovedAfterDelete) {
+          cleanupFailures.push(new Error("exact sandbox registry entry remains after cleanup"));
+        }
+      } catch (error) {
+        cleanupFailures.push(error);
+      }
+    }
+    try {
+      if (!fs.existsSync(stopPath)) fs.writeFileSync(stopPath, "stop\n", "utf8");
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts]: Fixed in 03d6240f: stop-marker creation now uses an atomic exclusive write and ignores only EEXIST.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 2]
```diff
@@ -0,0 +1,1375 @@
...
+    `agent_cwd = ${tomlString(input.shareDirectory)}`,
+    "agent_command = [",
+    `  ${tomlString(input.agentPath)},`,
+    `  ${tomlString(path.join(input.shareDirectory, "probe-agent.mjs"))},`,
+    "]",
+    "agent_env = [",
+    ...AGENT_ENVIRONMENT_NAMES.map((name) => `  ${JSON.stringify(name)},`),
+    "]",
+    "pc_least_privilege = false",
+    "pc_capabilities = []",
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts]: Addressed in 03d6240f: the qualification now requests pc_least_privilege = true, records that posture in the receipt, and documents it.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 3]
```diff
@@ -0,0 +1,1375 @@
...
+    "--auth",
+    "token",
+    "--bind",
+    "loopback",
+    "--port",
+    port,
+  ],
+  { env, stdio: "ignore", windowsHide: true },
+);
+writeFileSync(openClawPidPath, String(gateway.pid), "utf8");
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts]: Fixed in 03d6240f: the probe handles the spawn error event, writes the PID only when defined, records the spawn error, and emits a terminal outcome so host failure is prompt.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 4]
```diff
@@ -0,0 +1,1375 @@
...
+    ],
+    environment,
+    progress,
+    activityLabel,
+  );
+  if (result.exitCode === 3 || !result.stdout.trim()) return null;
+  if (result.exitCode !== 0) {
+    throw new Error(`host process identity query failed: ${commandDetail(result)}`);
+  }
+  return parseWindowsProcessIdentity(result.stdout);
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts]: Fixed in 03d6240f: exit code 3 is the only absent-process result; all other nonzero query results fail closed before stdout parsing. Added a regression test.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 5]
```diff
@@ -0,0 +1,1375 @@
...
+  const heartbeatPath = path.join(shareDirectory, "heartbeat.txt");
+  const openClawPidPath = path.join(shareDirectory, "openclaw.pid");
+  const stopPath = path.join(shareDirectory, "stop.txt");
+  const denyPath = path.join(runRoot, "denied-write.txt");
+  const gatewayLogPath = path.join(runRoot, "openshell-gateway.log");
+  const gatewayErrorPath = path.join(runRoot, "openshell-gateway.err.log");
+  const gatewayPort = await freeLoopbackPort();
+  const openClawPort = 20_000 + (Number.parseInt(runId.slice(0, 4), 16) % 20_000);
+  const sandboxName = `mxc-oc-${runId}`;
+  const gatewayName = `mxc-oc-${runId}`;
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts]: Fixed in 03d6240f: both ports use the loopback allocator, and sandbox and gateway names now use distinct prefixes.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 6]
```diff
@@ -0,0 +1,1375 @@
...
+    OPENSHELL_DRIVERS: "mxc",
+    OPENSHELL_GATEWAY: "",
+    OPENSHELL_GATEWAY_CONFIG: gatewayConfigPath,
+    XDG_CONFIG_HOME: configDirectory,
+    XDG_STATE_HOME: stateDirectory,
+  };
+  const controlEnvironment: NodeJS.ProcessEnv = {
+    ...gatewayEnvironment,
+    NEMOCLAW_MXC_E2E_TOKEN: undefined,
+  };
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts:1045]: Fixed in 03d6240f: host child processes now receive only an explicit Windows runtime allowlist plus the qualification variables. Arbitrary credential variables are excluded by test.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 7]
```diff
@@ -0,0 +1,1375 @@
...
+    checks = {
+      ...checks,
+      sandboxCreateAccepted:
+        create.exitCode === 0 ||
+        (checks.openClawHealth &&
+          checks.registryPresentWhileReady &&
+          !/connection refused|failed to connect|not registered|transport error/iu.test(
+            commandDetail(create),
+          )),
+    };
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts:1219]: Fixed in 03d6240f: sandbox creation is accepted only when the pinned OpenShell CLI returns exit code 0; the error-text denylist was removed.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 8]
```diff
@@ -0,0 +1,1375 @@
...
+        gateway.kill();
+        await Promise.race([
+          new Promise((resolve) => gateway?.once("exit", resolve)),
+          new Promise((resolve) => setTimeout(resolve, 5000)),
+        ]);
+      }
+      gatewayStopped = gateway === null || gateway.exitCode !== null;
+    } catch (error) {
+      cleanupFailures.push(error);
+    }
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts:1411]: Fixed in 03d6240f: the gateway PID is bound to its exact executable, creation identity, ordered port arguments, and TLS flag before a trusted taskkill fallback. Use of that fallback is recorded and fails qualification.

File: test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts [hunk 9]
```diff
@@ -0,0 +1,1375 @@
...
+      sensitiveRuntimeArtifactsRemoved = [
+        homeDirectory,
+        configDirectory,
+        stateDirectory,
+        gatewayLogPath,
+        gatewayErrorPath,
+      ].every((sensitivePath) => !fs.existsSync(sensitivePath));
+    } catch (error) {
+      cleanupFailures.push(error);
+    }
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container-helpers.ts]: Fixed in 03d6240f: descriptor closes and every sensitive-path removal are attempted independently, failures are collected, and removal status is recomputed afterward.

File: test/e2e/live/windows-mxc-openclaw-process-container.test.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  async ({ progress }) => {
+    progress.phase("qualify the Windows host and validate exact artifact identities");
+    const inputs = parseWindowsMxcOpenClawQualificationEnvironment(process.env);
+
+    progress.phase("start OpenClaw and verify in-sandbox readiness plus filesystem enforcement");
+    const receipt = await runWindowsMxcOpenClawProcessContainerQualification(inputs, progress);
+
+    progress.phase("delete the sandbox and verify registry plus OpenClaw process cleanup");
+    expect(receipt.verdict).toBe("pass");
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/live/windows-mxc-openclaw-process-container.test.ts:32]: Fixed in 03d6240f: the qualification helper enters the delete/registry/process-cleanup phase immediately before that lifecycle work, rather than after returning.

File: test/e2e/support/windows-mxc-openclaw-process-container.test.ts [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    ).toThrow(/must be clean/u);
+    expect(() =>
+      assertCleanCheckoutIdentity({
+        expectedRevision: "a".repeat(40),
+        observedRevision: "b".repeat(40),
+        statusOutput: "",
+      }),
+    ).toThrow(/does not match/u);
+    expect(normalizeReportedVersion("OpenClaw 2026.7.1\n")).toBe("2026.7.1");
+    expect(normalizeReportedVersion("2026.7.10\n")).not.toBe("2026.7.1");
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/support/windows-mxc-openclaw-process-container.test.ts]: Fixed in 03d6240f: the test now asserts the exact normalized 2026.7.10 value.

File: test/e2e/support/windows-mxc-openclaw-process-container.test.ts [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+    expect(createIndex).toBeGreaterThan(-1);
+    expect(firstDeleteIndex).toBeGreaterThan(createIndex);
+    expect(source).toContain("windows-mxc-sandbox-delete-cleanup");
+    expect(source).toContain('"taskkill.exe"');
+    expect(source).toContain('["System32", "taskkill.exe"]');
+    expect(source).toContain("sameWindowsProcessIdentity(identity, observed)");
+    expect(source.indexOf("fs.rmSync(runRoot")).toBeLessThan(
+      source.indexOf("const receipt: WindowsMxcOpenClawQualificationReceipt"),
+    );
+  });
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/support/windows-mxc-openclaw-process-container.test.ts]: Fixed in 03d6240f: the formatting-sensitive source assertions were replaced by TypeScript AST rules integrated into the semantic E2E checker, with formatting-independent positive and negative tests.

File: test/e2e/support/windows-mxc-openclaw-process-container.test.ts [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+        port: 17670,
+      }),
+    ).not.toThrow();
+    expect(() =>
+      assertExpectedOpenShellGatewayProcessIdentity(
+        { ...identity, commandLine: `${identity.commandLine} --other 9999` },
+        { gatewayPath: "C:\\package\\openshell-gateway.exe", port: 9999 },
+      ),
+    ).toThrow(/does not match/u);
+  });
```
[the remark this answers is not in the document]
Comment (senthilr-nv) [reply test/e2e/support/windows-mxc-openclaw-process-container.test.ts:246]: Fixed in 0bce964a: the negative fixture now keeps expected port 17670 while placing 17670 before --port, so both tokens are present but the ordered pair is invalid. The focused support suite passes.

**Comment by senthilr-nv** [comment]
Addressed the latest review feedback in `7b1f42b2`:

- fail immediately on a nonzero `sandbox create` result and preserve its command detail;
- revalidate pinned OpenShell, `wxc-exec`, Node.js, and OpenClaw artifacts immediately before security-relevant use, with a replacement regression test;
- fail loudly if the path-bound MXC semantic guard disappears, and cover missing lifecycle commands;
- clarify that delete retry and process termination are failure containment and cannot produce a passing qualification.

Focused E2E support tests pass (35/35), semantic phase coverage passes (118 tests across 75 files), repository checks and docs validation pass, and the exact-head security and documentation reviews pass.

**Review by apurvvkumaria — APPROVED** [review]

**Review by cjagwani — COMMENTED** [review]
Security review — exact head `afcd6266ce6af736a9018bcd19c68b6afabe08af` against base `cbabb66bfc985f3474714f95bfbc6ec1992441f9`: **PASS with no findings**.

1. **Secrets and credentials — PASS.** The opt-in target creates an ephemeral random readiness token, keeps it out of arguments and receipts, redacts command output, and passes host children only an explicit Windows runtime-variable allowlist. No repository secret or reusable credential is added.
2. **Input validation and injection resistance — PASS.** Required inputs are validated as exact revisions, versions, paths, hashes, ports, and process identities. Runtime commands use argument arrays rather than shell interpolation. Files and artifact trees are resolved, type-checked, bounded, and rejected when they contain links or unsupported entries.
3. **Authentication and authorization — PASS.** The temporary gateway is loopback-bound and token-protected. The target neither registers nor activates MXC as a supported runtime and does not widen repository or host permissions.
4. **Dependencies and supply chain — PASS.** No dependency or lockfile changes. NemoClaw, OpenShell, `wxc-exec`, Node, OpenClaw entry, and the complete OpenClaw artifact tree are identity-pinned and revalidated immediately before security-relevant use.
5. **Error handling and information exposure — PASS.** Create, query, identity, policy, and cleanup failures are fail-closed. Child output is size-bounded, diagnostics omit the token, and the qualification receipt contains only non-secret evidence.
6. **Cryptography and data protection — PASS.** `randomBytes(32)` supplies the ephemeral token. SHA-256 is used for deterministic artifact identity, not as an authentication substitute. Sensitive temporary state is removed during terminal cleanup.
7. **Configuration and infrastructure — PASS.** The generated gateway and filesystem policy are temporary and least-privilege; network access is limited to loopback. The qualification is inactive and opt-in, with bounded processes, timeouts, output, artifact traversal, and cleanup scope.
8. **Security testing — PASS.** Support contracts cover identity drift and replacement, dirty-source and version-prefix rejection, complete process identity, token exclusion, environment filtering, tree-content/path/link/device cases, semantic phases, policy allow/deny behavior, and cleanup fallbacks. Exact GitHub Actions and protected qualification evidence remain mandatory.
9. **System security — PASS.** Emergency process termination occurs only after re-querying and matching the recorded executable, arguments, parent/child relationship, and port identity. Cleanup targets exact generated names and run directories and reports any residual sandbox, process, registry, or sensitive path.

The current head is a signed mechanical merge of current `main`. Its effective eight-file patch is byte-for-byte unchanged from reviewed head `98adf64d2b1c7d5275c801dadc30c9c44ff34911` (raw patch SHA-256 `285c9a4cd764c51605f45f8a8bdaa9e557ff6db3d91980bd62cf80268b779734`; stable patch ID `d6bf3bd8aef510d5f59dc5c5d803216d62c0d3df`). No product-scope waiver or gate waiver is used.

**Review by cjagwani — COMMENTED** [review]
Security review — `99e54157678400cfda425bdc9a1f0e531749e6ea` against base `cbabb66bfc985f3474714f95bfbc6ec1992441f9`: **PASS with no findings**.

The nine-category review at parent `afcd6266ce6af736a9018bcd19c68b6afabe08af` is carried forward without expansion. The only new change is a documentation correction in `test/e2e/README.md`: it now says a secret-free receipt is written for either verdict only after preflight and local setup succeed. That matches the implementation and removes an unsafe expectation that early identity or host-validation failures would always leave a receipt.

1. Secrets and credentials — PASS: no secret-bearing behavior or example changes.
2. Input validation and injection resistance — PASS: no executable path changes; exact identity and argument-array controls remain intact.
3. Authentication and authorization — PASS: the loopback token boundary and inactive/no-support scope are unchanged.
4. Dependencies and supply chain — PASS: no dependency or artifact source changes.
5. Error handling and information exposure — PASS: the documentation now accurately distinguishes pre-receipt failures from verdict-bearing runs.
6. Cryptography and data protection — PASS: token generation, SHA-256 identity checks, and sensitive-state cleanup are unchanged.
7. Configuration and infrastructure — PASS: no policy, privilege, host, network, or workflow change.
8. Security testing — PASS: the previous exact eight-file security contracts remain unchanged; Markdown lint and the repository test-file budget passed for the follow-up.
9. System security — PASS: no process, cleanup, filesystem, or trusted-boundary behavior changes.

Commit `99e54157678400cfda425bdc9a1f0e531749e6ea` is signed and Signed-off-by. Fresh CI, independent approval, and protected E2E remain mandatory; no prior-head result is accepted.

**Comment by cjagwani** [comment]
The required documentation-writer review found and fixed one exact operator-contract issue in signed, GitHub-Verified commit `99e54157678400cfda425bdc9a1f0e531749e6ea`: receipts are written for either verdict only after preflight and local setup succeed; earlier validation failures may occur before receipt creation. Normal Markdown lint and the repository test-file budget passed. The exact-head nine-category security review passed, and fresh CI plus protected E2E are required. Please review the current head so the post-push approval rule can be satisfied.
